src/app/static/src/app/components/genericChart/GenericChart.vue

```diff
@@ -101,34 +144,92 @@
+                        ...subplots,
+                        rows
+                    };
+                    //Height per row plus enough to accommodate margin and padding
+                    scrollHeight = `${(subplots.heightPerRow * rows) + 70}px`;
+                }
+
+                // The metadata supports multiple chart types per chart e.g Scatter and Bar, but for now we only need to
+                // support one chart type, so here we select the first config in the array
+                const chartConfig = this.chartMetadata.chartConfig[0].config;
+
                 return {
-                    dataSourceConfigValues
+                    dataSourceConfigValues,
+                    layoutData,
+                    scrollHeight,
+                    chartConfig
                 };
+            },
+            chartData() {
+                const result = {} as Dict<unknown[]>;
+
+                for (const dataSource of this.chartMetadata.dataSelectors.dataSources) {
+                    const dataSourceId = dataSource.id;
+                    const dataSourceSelections = this.dataSourceSelections[dataSourceId];
+                    const datasetId = dataSourceSelections.datasetId;
+                    const unfilteredData = this.datasets[datasetId]?.data;
+
+                    const filters = this.datasets[dataSourceSelections.datasetId]?.metadata.filters || [];
+                    const selectedFilterOptions = dataSourceSelections.selectedFilterOptions;
+
+                    if (!unfilteredData || !selectedFilterOptions) {
+                        return null; //Do not attempt to initialise if we are missing any datasets, or selections not initialised
+                    }
+
+                    result[dataSourceId] = filterData(unfilteredData, filters, selectedFilterOptions);
+                }
+                return result;
             }
         },
         methods: {
             getDataset: mapActionByName(namespace, 'getDataset'),
-            ensureDataset(datasetId: string) {
+            async ensureDataset(datasetId: string) {
                 if (datasetId && !this.datasets[datasetId]) {
-                    const dataset = this.chartMetadata.datasets.find(dataset => dataset.id === datasetId)!;
-                    this.getDataset({datasetId, url: dataset.url});
+                    const datasetConfig = this.chartMetadata.datasets.find(dataset => dataset.id === datasetId)!;
+                    await this.getDataset({datasetId, url: datasetConfig.url});
                 }
             },
-            updateDataSource(dataSourceId: string, datasetId: string) {
+            async updateDataSource(dataSourceId: string, datasetId: string) {
                 this.dataSourceSelections[dataSourceId]!.datasetId = datasetId;
-                this.ensureDataset(datasetId);
+                await this.ensureDataset(datasetId);
+                this.setDataSourceDefaultFilterSelections(dataSourceId, datasetId);
```

Comment:
As observed by others, there is a tricky bug here: L214 makes a change to state that is depended on by `chartData()`. L215 is essentially a break in control flow, meaning that the change can propagate all the way through to the chart re-rendering before L216 is executed. If this happens the new chart data is filtered using the old filters (e.g. art data with "anc_clients" plot type)... resulting in a zero length `data` array. Swapping the order of lines 214 and 215 is sufficient to fix this, but it would be preferable to restructure things so that state updates are performed more locally/atomically.

Reply:
Yes, good spot, thanks @mwoodbri - in fact, it would probably be good to also delay setting the new datasource id as a selection until we've set its default filter selections. I'll make sure that we see the spinner while the fetch is happening too.